pkg/snapshot/snapshotter/snapshotter.go

```diff
@@ -242,9 +242,20 @@ func (ssr *Snapshotter) stop() {
 		ssr.deltaSnapshotTimer = nil
 	}
 	ssr.closeEtcdClient()
+}
 
+// SetSnapshotterInactive set the snapshotter state to Inactive.
+func (ssr *Snapshotter) SetSnapshotterInactive() {
```

Comment:
So there are a few things that should be corrected:

**Method:** snapshotter.snapshotEventHandler
```go
defer leaseUpdateCancel()
...
case <-stopCh:
	ssr.logger.Info("Closing the Snapshot EventHandler.")
	leaseUpdateCancel()
	ssr.cleanupInMemoryEvents()
	return nil
}
```
function `leaseUpdateCancel` is called twice. Not necessary.
2. So consider the following:
    1. Backup-restore loses leadership. As a result the leCtx gets cancelled. In method `handleSsrStopRequest` an empty struct will be sent over the ssr stop channel.
    2. As a result ssr.snapshotEventHandler will exit and the status of the ssr will still be active.
    3. Now ssr.TriggerFullSnapshot is invoked. It will be able to take the lock as currently no one is holding it and its status would still be active. It will now send a message on `ssr.fullSnapshotAckCh` with no receiver. This will block.

Reply:
> function leaseUpdateCancel is called twice. Not necessary.

removed

> * Backup-restore loses leadership. As a result the leCtx gets cancelled. In method handleSsrStopRequest an empty struct will be sent over the ssr stop channel.
> * As a result ssr.snapshotEventHandler will exit and the status of the ssr will still be active.
 >* Now ssr.TriggerFullSnapshot is invoked. It will be able to take the lock as currently no one is holding it and its status would still be active. It will now send a message on ssr.fullSnapshotReqCh with no receiver. This will block. Please double check if this can happen.

as first safety measure I have removed the invokation of `ssr.TriggerFullSnapshot` after defrag call.
as secondary safety measure, I'm setting the snapshottter state to `Inactive` before exiting snapshotEventHandler loop.

Reply:
> It is also incorrect to create a channel in OnStartedLeading callback. It should be created as part of snapshotter.NewSnapshotter call. The channel belongs to snapshotter and should not be created outside.

but we are passing the responsibility of closing the snapshottter to `handleSsrStopRequest`, that's why we are creating channel in OnStartedLeading.